devito/archinfo.py

```diff
@@ -60,16 +61,39 @@ def get_cpu_brand():
             except:
                 pass
 
-            return None
+            try:
+                # Certain ARM CPUs, e.g. Marvell Thunder X2
+                return cpuinfo.get_cpu_info().get('arch').lower()
+            except:
+                return None
 
         cpu_info['flags'] = get_cpu_flags()
         cpu_info['brand'] = get_cpu_brand()
 
-    if not all(i in cpu_info for i in ('flags', 'brand')):
-        # Fallback
-        ci = cpuinfo.get_cpu_info()
-        cpu_info['flags'] = ci.get('flags')
-        cpu_info['brand'] = ci.get('brand')
+    try:
+        if cpu_info['flags'] is None:
+            cpu_info['flags'] = cpuinfo.get_cpu_info().get('flags')
+    except KeyError:
+        cpu_info['flags'] = cpuinfo.get_cpu_info().get('flags')
+
+    try:
+        if cpu_info['brand'] is None:
+            cpu_info['brand'] = cpuinfo.get_cpu_info().get('brand')
+    except KeyError:
+        cpu_info['brand'] = cpuinfo.get_cpu_info().get('brand')
+
+    # Detect number of physical and logical cores
+    try:
+        if 'arm' in cpu_info['brand']:
+            # Special case: In some ARM processors psutils fail to detect physical cores
+            # correctly so we use lscpu()
+            logical = psutil.cpu_count(logical=True)
+            physical = (lscpu()['Core(s) per socket'] * lscpu()['Socket(s)'])
+            cpu_info['logical'] = logical
```

Comment:
I still think that the issue with ARM is the correct detection of number of physical cores, **not** the logical cores.  So I would:

* drop the `logical` detection here
* move this try-except after the "logical core detection" part, ie below line 110

Reply:
Yes, but getting the logical here helps as return `cpu_info` and avoid all the other lines after 110

Reply:
I was suggesting to move it right between lines 103 and 104, so you would still avoid "those other lines"

Reply:
if we do so, I think, it'd become a bit neater